sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkSpanBackend.scala

```diff
@@ -0,0 +1,277 @@
+      _ <- end(now)
+    } yield ()
+
+  private[otel4s] def end(timestamp: FiniteDuration): F[Unit] =
+    mutableState.update { s =>
+      if (s.hasEnded) s // todo: log warn
+      else s.copy(endEpochNanos = timestamp.toNanos, hasEnded = true)
+    } >> spanProcessor.onEnd(this)
+
+  private def addTimedEvent(event: EventData): F[Unit] =
+    mutableState.update { s =>
+      if (s.events.sizeIs <= spanLimits.maxNumberOfEvents) {
+        s.copy(
+          events = s.events :+ event,
+          totalRecordedEvents = s.totalRecordedEvents + 1
+        )
+      } else {
+        s.copy(totalRecordedEvents = s.totalRecordedEvents + 1)
+      }
+    }
+
+  def name: String =
+    immutableState.name // todo: name can be mutated by internals
+
+  def kind: SpanKind =
+    immutableState.kind
+
+  def scopeInfo: InstrumentationScopeInfo =
+    immutableState.scopeInfo
+
+  def spanContext: SpanContext =
+    immutableState.context
+
+  def toSpanData: F[SpanData] =
+    for {
+      state <- mutableState.get
+    } yield SpanData.create(
+      name = immutableState.name,
+      kind = immutableState.kind,
+      spanContext = immutableState.context,
+      parentSpanContext = immutableState.parentContext,
+      status = state.status,
+      startEpochNanos = immutableState.startEpochNanos,
+      attributes = state.attributes,
+      events = state.events,
+      links = immutableState.links,
+      endEpochNanos = state.endEpochNanos,
+      hasEnded = state.hasEnded,
+      totalRecordedEvents = state.totalRecordedEvents,
+      totalRecordedLinks = immutableState.totalRecordedLinks,
+      totalAttributeCount =
+        state.attributes.size, // todo: incorrect when limits are applied,
+      instrumentationScopeInfo = immutableState.scopeInfo,
+      resource = immutableState.resource
+    )
+
+  def hasEnded: F[Boolean] =
+    mutableState.get.map(_.hasEnded)
+
+  def latencyNanos: F[Long] =
+    for {
+      state <- mutableState.get
+      endEpochNanos <-
+        if (state.hasEnded) Monad[F].pure(state.endEpochNanos)
+        else Clock[F].realTime.map(_.toNanos)
+    } yield endEpochNanos - immutableState.startEpochNanos
+
+  def getAttribute[A](key: AttributeKey[A]): F[Option[A]] =
+    for {
+      state <- mutableState.get
+    } yield state.attributes.get(key).map(_.value)
+}
+
+object SdkSpanBackend {
+
+  private final case class ImmutableState(
+      name: String,
+      context: SpanContext,
+      scopeInfo: InstrumentationScopeInfo,
+      kind: SpanKind,
+      parentContext: Option[SpanContext],
+      resource: Resource,
+      links: List[LinkData],
+      totalRecordedLinks: Int,
+      startEpochNanos: Long
+  )
+
+  private final case class MutableState(
+      status: StatusData,
+      attributes: Attributes,
+      events: List[EventData],
+      totalRecordedEvents: Int,
+      endEpochNanos: Long,
+      hasEnded: Boolean
+  )
+
+  // todo: replace Concurrent with Async
+  def start[F[_]: Concurrent: Clock](
```

Comment:
`: Concurrent: Clock` - can be replaced with `Temporal` if we decide to always rely on the CE Clock

Reply:
Use `Temporal` 👍 there is no reason to mess with `Clock` independently.